igcollect/sfp_dom.py

```diff
@@ -0,0 +1,372 @@
+#!/usr/bin/env python3
+"""igcollect - SFP Digital Optical Monitoring metric collector
+
+Copyright (c) 2022 InnoGames GmbH
+
+This script reads SFP EEPROM and parses it in order to collect optical
+diagnostics data.
+
+To properly understand what happens on this script, a read of the SFF-8472
+standard is highly recommended.
+
+In some places, we gave up on making things Pythonic in detriment of shaving
+off a few ms on execution time.
+"""
+import ctypes
+import fcntl
+import socket
+import sys
+
+from argparse import ArgumentParser
+from os import listdir
+from os.path import isdir, islink, join
+from time import time
+
+IFNAMSIZ = 16
+
+SIOCETHTOOL = 0x8946
+
+ETHTOOL_GEEPROM = 0x0000000b  # Get EEPROM data
+ETHTOOL_GMODULEINFO = 0x00000042  # Get plug-in module information
+ETHTOOL_GMODULEEEPROM = 0x00000043  # Get plug-in module EEPROM
+
+ETH_MODULE_SFF_8079 = 0x1
+ETH_MODULE_SFF_8079_LEN = 256
+ETH_MODULE_SFF_8472 = 0x2
+ETH_MODULE_SFF_8472_LEN = 512
+ETH_MODULE_SFF_8636 = 0x3
+ETH_MODULE_SFF_8636_LEN = 256
+ETH_MODULE_SFF_8436 = 0x4
+ETH_MODULE_SFF_8436_LEN = 256
+
+SFF_A0_DOM = 92
+SFF_A0_DOM_PWRT = (1 << 3)
+SFF_A0_DOM_EXTCAL = (1 << 4)
+
+SFF_A2_BASE = 0x100
+SFF_A2_TEMP = 96
+SFF_A2_TEMP_HALRM = 0
+SFF_A2_TEMP_LALRM = 2
+SFF_A2_TEMP_HWARN = 4
+SFF_A2_TEMP_LWARN = 6
+SFF_A2_VCC = 98
+SFF_A2_VCC_HALRM = 8
+SFF_A2_VCC_LALRM = 10
+SFF_A2_VCC_HWARN = 12
+SFF_A2_VCC_LWARN = 14
+SFF_A2_BIAS = 100
+SFF_A2_BIAS_HALRM = 16
+SFF_A2_BIAS_LALRM = 18
+SFF_A2_BIAS_HWARN = 20
+SFF_A2_BIAS_LWARN = 22
+SFF_A2_TX_PWR = 102
+SFF_A2_TX_PWR_HALRM = 24
+SFF_A2_TX_PWR_LALRM = 26
+SFF_A2_TX_PWR_HWARN = 28
+SFF_A2_TX_PWR_LWARN = 30
+SFF_A2_RX_PWR = 104
+SFF_A2_RX_PWR_HALRM = 32
+SFF_A2_RX_PWR_LALRM = 34
+SFF_A2_RX_PWR_HWARN = 36
+SFF_A2_RX_PWR_LWARN = 38
+
+
+class ethtool_modinfo(ctypes.Structure):
+    _fields_ = [
+        ('cmd', ctypes.c_uint32),
+        ('type', ctypes.c_uint32),
+        ('eeprom_len', ctypes.c_uint32),
+        ('reserved', ctypes.c_uint32 * 8)
+    ]
+
+
+class ethtool_eeprom(ctypes.Structure):
+    _fields_ = [
+        ('cmd', ctypes.c_uint32),
+        ('magic', ctypes.c_uint32),
+        ('offset', ctypes.c_uint32),
+        ('len', ctypes.c_uint32),
+        ('data', ctypes.c_byte * 512)
+    ]
+
+
+class ifr_data(ctypes.Union):
+    _fields_ = [
+        ('ethtool_modinfo_ptr', ctypes.POINTER(ethtool_modinfo)),
+        ('ethtool_eeprom_ptr', ctypes.POINTER(ethtool_eeprom)),
+    ]
+
+
+class ifreq(ctypes.Structure):
+    _fields_ = [
+        ('ifr_name', (ctypes.c_ubyte * IFNAMSIZ)),
+        ('ifr_data', ifr_data)
+    ]
+
+
+class SFPData:
+    def __init__(self, data, interface):
+        self._data = data
+        self.interface = interface
+        self.metrics = {
```

Comment:
you could make this a constant on the top, something like this:
```
METRICS_MAP = {
    'temperature': {
        'current': SFF_A2_TEMP,
        'low_warning': SFF_A2_TEMP_LWARN,
        ...
    },
    'bias_current': {
        'current': SFF_A2_BIAS,
        ...
    },
    ...
}
```
also you can define another fn map like so:
```
FN_MAP = {
    'temperature': a2_offset_to_celsius,
    ...
}
```

that way you can do the following:
- initialize your empty metrics dict here with dict comprehension
- map the right conversion functions to the right data points

however you would need to slightly rearrange some stuff so you can actually reference it

Reply:
I decided to include the maps in the SFPData class. It made more sense to put them there, especially because the FN_MAP needed the functions to be declared beforehand.

Reply:
I second this. Was already thinking about that while reading through the code. The different calculation functions were already screaming for that in my mind. ^^ It would go in the direction of SOLID-O (so, open for extension / closed for modification), which would mean that `SFPData` or at least the method `SFPData::_parse_eeprom_data` would be open for extension while being closed for modification. Meaning, that to add more functionality, you could do that then without touching the function (or class) at all. This would make it easier and faster to maintain.

Maybe another idea could be to not have 2 configuration points:
``` python
METRICS_MAP = {
    'temperature': {
        'current': (SFF_A2_TEMP, a2_offset_to_celsius),
        'low_warning': (SFF_A2_TEMP_LWARN, a2_offset_to_celsius),
        ...
    },
    'bias_current': {
        'current': (SFF_A2_BIAS, a2_offset_to_milliamps),
        ...
    },
    ...
}
```

You could use that then like that:
``` python
def _parse_eeprom_data(self):
    ...
    for category, sub_metrics in METRICS_MAP.items():
        for metric, (offset, fn) in sub_metrics.items():
            self.metrics[category][metric] = fn(self._data, offset)
```

Reply:
This might be a bit slower though than the direct usage of the functions.

Reply:
The functions differ per the outer key of the `METRICS_MAP` dict. 
I feel like adding the function on each inner key is too redundant. 

I thought about:
```python
METRICS_MAP = {
    'temperature': {
        'function': a2_offset_to_celsius,
        'current': SFF_A2_TEMP,
        'low_warning': SFF_A2_TEMP_LWARN,
        ...
    },
    'bias_current': {
        'function': a2_offset_to_milliamps,
        'current': SFF_A2_BIAS,
        ...
    },
    ...
}
Since we iterate over `self.metrics`, it should be fine key-wise.

Reply:
I would actually prefer to have just one point of configuration. So, we won't have duplicated configuration points. This would make it easier to maintain.

Your suggestion would be a bit tricky, as the you put the `function` on the same level as the `metric` names. To avoid that, I would suggest to either:
``` python
METRICS_MAP = {
    'temperature': {
        '__function': a2_offset_to_celsius,
        'current': SFF_A2_TEMP,
        'low_warning': SFF_A2_TEMP_LWARN,
        ...
    },
    'bias_current': {
        '__function': a2_offset_to_milliamps,
        'current': SFF_A2_BIAS,
        ...
    },
    ...
}
```

or

``` python
METRICS_MAP = {
    'temperature': {
        'function': a2_offset_to_celsius,
        'metrics': {
            'current': SFF_A2_TEMP,
            'low_warning': SFF_A2_TEMP_LWARN,
            ...
        }
    },
    'bias_current': {
        'function': a2_offset_to_milliamps,
        'metrics': {
            'current': SFF_A2_BIAS,
            ...
        }
    },
    ...
}
```

Reply:
I will send a second PR soon extending some of the functionality and adding a new type of data structure that will address this.
There were quite some changes around this part, so I'd rather leave it as is for this PR. I took these comments into consideration when writing the new set of patches.